### PR TITLE
Force upgrade pods of a same StatefulSet

### PR DESCRIPTION
**Pull request by sebgl**
Prior to that change, we trigger a force upgrade for Pods that are
either Pending or bootlooping, if all Pods in the cluster are also
Pending or bootlooping.
A situation where Pods of the first StatefulSet are fine but Pods of the
2nd StatefulSet are bootlooping can still happen.

This fixes it by checking the state of all Pods of the same StatefulSet,
instead of considering all Pods in the cluster.

Fixes https://github.com/elastic/cloud-on-k8s/issues/1847.

### CODE REVIEW

**Comment by barkbay** [comment]
I'm not sure it is a fix for #1847 as this PR does not seem to take into account that the API call may return 503 and prevents the pods to be force upgraded.

File: pkg/controller/elasticsearch/driver/upgrade_forced.go [hunk 1]
```diff
@@ -5,37 +5,68 @@
...
+func podsByStatefulSetName(pods []corev1.Pod) map[string][]corev1.Pod {
+	byStatefulSet := map[string][]corev1.Pod{}
+	for _, p := range pods {
+		ssetName, exists := p.Labels[label.StatefulSetNameLabelName]
+		if !exists {
+			log.Error(
+				fmt.Errorf("expected label %s not set", label.StatefulSetNameLabelName),
+				"skipping forced upgrade",
+				"namespace", p.Namespace, "pod_name", p.Name)
+			continue
```
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/upgrade_forced.go]: So if one pod is missing a label we are just dropping that pod from the algorithm and logging an error? I wonder how we could get into such a situation? The only thing I can think of is a user manually deleting a label.

**Comment by sebgl** [comment]
> I'm not sure it is a fix for #1847 as this PR does not seem to take into account that the API call may return 503 and prevents the pods to be force upgraded.

The force upgrade will happen whether or not the API returns a 503.

File: pkg/controller/elasticsearch/driver/upgrade_forced.go (hunk 1 above)
Comment (sebgl) [reply pkg/controller/elasticsearch/driver/upgrade_forced.go]: Yeah that should definitely not happen :/
Not sure what more I can do here.

**Comment by barkbay** [comment]
> > I'm not sure it is a fix for #1847 as this PR does not seem to take into account that the API call may return 503 and prevents the pods to be force upgraded.
> 
> The force upgrade will happen whether or not the API returns a 503.

I don't understand how it is possible if an error is raised [here](https://github.com/elastic/cloud-on-k8s/blob/c07ddeff921edc7f4bdf4a23c04a7f0f5399af2d/pkg/controller/elasticsearch/driver/nodes.go#L92) ?

**Comment by sebgl** [comment]
> I don't understand how it is possible if an error is raised here 
You're right @barkbay! This only fixes cases where:

* es is not reachable
* es appears and is reachable

but not:
* es appears but is not reachable

**Comment by barkbay** [comment]
No worries!
It is still great you fixed these ones 👍

**Pull request merged**